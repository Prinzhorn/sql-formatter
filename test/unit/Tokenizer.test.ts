import Tokenizer from 'src/lexer/Tokenizer';

describe('Tokenizer', () => {
  const tokenize = (sql: string) =>
    new Tokenizer({
      reservedCommands: ['FROM', 'WHERE', 'LIMIT', 'CREATE TABLE'],
      reservedSelect: ['SELECT'],
      reservedDependentClauses: ['WHEN', 'ELSE'],
      reservedSetOperations: ['UNION', 'UNION ALL'],
      reservedJoins: ['JOIN'],
      reservedFunctionNames: ['SQRT', 'CURRENT_TIME'],
      reservedKeywords: ['BETWEEN', 'LIKE', 'ON', 'USING'],
      stringTypes: ["''-qq"],
      identTypes: ['""-qq'],
    }).tokenize(sql, {});

  it('tokenizes whitespace to empty array', () => {
    expect(tokenize(' \t\n \n\r ')).toEqual([]);
  });

  it('tokenizes single line SQL tokens', () => {
    expect(tokenize('SELECT * FROM foo;')).toMatchSnapshot();
  });

  it('tokenizes multiline SQL tokens', () => {
    expect(tokenize('SELECT "foo\n bar" /* \n\n\n */;')).toMatchSnapshot();
  });
});
