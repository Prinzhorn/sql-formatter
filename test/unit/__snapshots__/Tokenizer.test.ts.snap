// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer tokenizes multiline SQL tokens 1`] = `
Array [
  Object {
    "loc": Object {
      "end": 6,
      "start": 0,
    },
    "precedingWhitespace": undefined,
    "raw": "SELECT",
    "text": "SELECT",
    "type": "RESERVED_SELECT",
  },
  Object {
    "loc": Object {
      "end": 17,
      "start": 7,
    },
    "precedingWhitespace": " ",
    "raw": "\\"foo
 bar\\"",
    "text": "\\"foo
 bar\\"",
    "type": "QUOTED_IDENTIFIER",
  },
  Object {
    "loc": Object {
      "end": 27,
      "start": 18,
    },
    "precedingWhitespace": " ",
    "raw": "/* 


 */",
    "text": "/* 


 */",
    "type": "BLOCK_COMMENT",
  },
  Object {
    "loc": Object {
      "end": 28,
      "start": 27,
    },
    "precedingWhitespace": undefined,
    "raw": ";",
    "text": ";",
    "type": "DELIMITER",
  },
]
`;

exports[`Tokenizer tokenizes single line SQL tokens 1`] = `
Array [
  Object {
    "loc": Object {
      "end": 6,
      "start": 0,
    },
    "precedingWhitespace": undefined,
    "raw": "SELECT",
    "text": "SELECT",
    "type": "RESERVED_SELECT",
  },
  Object {
    "loc": Object {
      "end": 8,
      "start": 7,
    },
    "precedingWhitespace": " ",
    "raw": "*",
    "text": "*",
    "type": "ASTERISK",
  },
  Object {
    "loc": Object {
      "end": 13,
      "start": 9,
    },
    "precedingWhitespace": " ",
    "raw": "FROM",
    "text": "FROM",
    "type": "RESERVED_COMMAND",
  },
  Object {
    "loc": Object {
      "end": 17,
      "start": 14,
    },
    "precedingWhitespace": " ",
    "raw": "foo",
    "text": "foo",
    "type": "IDENTIFIER",
  },
  Object {
    "loc": Object {
      "end": 18,
      "start": 17,
    },
    "precedingWhitespace": undefined,
    "raw": ";",
    "text": ";",
    "type": "DELIMITER",
  },
]
`;
